### ISSUE TITLE
HTML converted toc2 link in the notebook cell redirects back to notebook

### ISSUE THREAD

**Issue by p2ya** [ipython-contrib/jupyter_contrib_nbextensions#1136]
used “t” link to insert a toc cell in the notebook.
Although, the side and cell toc links back to stated section in notebook itself, when converted to html through:
jupyter nbconvert --to html_embed --template toc2 notebook.ipynb
the toc on side works fine
BUT
the toc as notebook cell redirects to the notebook link not to the section as expected.
If possible I want the toc in notebook cell to link to notebook sections as well when converted to html.

**Comment by jcb91** [comment]
> the toc as notebook cell redirects to the notebook link

I don't really understand what you mean by "the notebook link". Do you mean that the link points to a notebook server?

**Comment by p2ya** [comment]
yes, when converted to html it redirects back to notebook server like: (http://localhost:8889/notebooks/notebook.ipynb#section of notebook)
however the side toc redirects to correct part of notebook in html.

**Comment by jcb91** [comment]
hmm, ok. This (getting links in live notebooks and static html, which may be embedded in other pages, to link to the right place) has been a right pain - see also #1083 (exact same issue), #1034, #1023

**Comment by p2ya** [comment]
referencing back to issue #1083, I checked the file and it has the following code:
var make_link = function(h, toc_mod_id) {
        var a = $('<a>')
            .attr({
                'href': h.find('.anchor-link').attr('href'),
                'data-toc-modified-id': toc_mod_id,
            });
from line 72-77

What do you recommend to be done to fix this issue ?

**Comment by jcb91** [comment]
That is the current source, and it produces links relative to the current page, as far as I'm aware. This means that the toc cell will contain links formed when the notebook is running, which will persist into the html source. However, the links _should_ get rerendered on loading the html in a browser. Are you looking at the html source, or loading the page in a browser? Because (I think!) in a browser, the cell should get rerendered on load...

**Comment by jcb91** [comment]
wait, given you mentioned the sidebar, I assume you are looking at a browser :thinking:

**Comment by p2ya** [comment]
I use:
jupyter nbconvert --to html_embed --template toc2 notebook.ipynb
then I download the html and open it in browser

**Comment by jcb91** [comment]
Hmm, ok. I'm not sure what's going on then, I'll have to think about it. Also worth pinging @jfbercher, although I think he's quite busy at the moment...

**Comment by p2ya** [comment]
Thanks, really appreciate to get updated on this and resolve it

**Comment by jcb91** [comment]
Ok, on inspection, I think the toc cell may be failing to update because the template is still explicitly setting `toc_cell = false`...

**Comment by jcb91** [comment]
see #1141 for an attempt to fix the template to get the cell to rerender in the browser...

**Comment by p2ya** [comment]
I have:
{
  "toc": "true"
}
in metadata of cell toc

**Comment by jcb91** [comment]
Yeah, it's not the cell's metadata that's relevant here, though, it's the settings which the template is using to load the nbextension javascript, in which `toc_cell = false` is hardcoded, since in the past we didn't attempt to reload cell toc in a non-live notebook. See [changes in #1141](https://github.com/ipython-contrib/jupyter_contrib_nbextensions/pull/1141/files) for details

**Comment by p2ya** [comment]
I tried to test it by adding:
var cfg = {{ nb.get('metadata', {}).get('toc', {})|tojson|safe }};
into my toc2.tpl
and restarted the Jupyter
howeve rit didnt work, am I missing something to make this work ?

**Comment by jcb91** [comment]
Not that I can think of, as long as you added it after/instead of the previous definition of cfg. I guess I'll have to do some more extensive testing...

**Comment by p2ya** [comment]
yes I added after 
saved and restarted Jupyter

**Comment by p2ya** [comment]
Hi @jcb91
Did you have a chance to identify the potential issue and fix ?
can you replicate the issue as well ?

**Comment by jcb91** [comment]
Hi @p2ya, unfortunately I'm actually unable to reproduce this issue at all on my local notebook server - the links in the output html are just the anchor element (the bit after `#`). What versions are you using of `jupyter_contrib_nbextensions`, `notebook` and `nbconvert`?

a bit of detail:

source notebook:

```
$ grep -e "a href" notebook.ipynb
    "<div class=\"toc\" style=\"margin-top: 1em;\"><ul class=\"toc-item\"><li><span><a href=\"#Main-section\" data-toc-modified-id=\"Main-section-1\"><span class=\"toc-item-num\">1&nbsp;&nbsp;</span>Main section</a></span><ul class=\"toc-item\"><li><span><a href=\"#subsection-a\" data-toc-modified-id=\"subsection-a-1.1\"><span class=\"toc-item-num\">1.1&nbsp;&nbsp;</span>subsection a</a></span></li><li><span><a href=\"#subsection-b\" data-toc-modified-id=\"subsection-b-1.2\"><span class=\"toc-item-num\">1.2&nbsp;&nbsp;</span>subsection b</a></span></li></ul></li><li><span><a href=\"#Second-main-section\" data-toc-modified-id=\"Second-main-section-2\"><span class=\"toc-item-num\">2&nbsp;&nbsp;</span>Second main section</a></span><ul class=\"toc-item\"><li><span><a href=\"#subsection-c\" data-toc-modified-id=\"subsection-c-2.1\"><span class=\"toc-item-num\">2.1&nbsp;&nbsp;</span>subsection c</a></span></li><li><span><a href=\"#subsection-d\" data-toc-modified-id=\"subsection-d-2.2\"><span class=\"toc-item-num\">2.2&nbsp;&nbsp;</span>subsection d</a></span></li></ul></li></ul></div>"
```
conversion:

```
$ jupyter nbconvert --to html_embed --template toc2 notebook.ipynb 
[NbConvertApp] Converting notebook notebook.ipynb to html_embed
[NbConvertApp] Writing 255025 bytes to notebook.html
```

output html:
```html
$ grep -e "a href" notebook.html 
<div class="toc" style="margin-top: 1em;"><ul class="toc-item"><li><span><a href="#Main-section" data-toc-modified-id="Main-section-1"><span class="toc-item-num">1  </span>Main section</a></span><ul class="toc-item"><li><span><a href="#subsection-a" data-toc-modified-id="subsection-a-1.1"><span class="toc-item-num">1.1  </span>subsection a</a></span></li><li><span><a href="#subsection-b" data-toc-modified-id="subsection-b-1.2"><span class="toc-item-num">1.2  </span>subsection b</a></span></li></ul></li><li><span><a href="#Second-main-section" data-toc-modified-id="Second-main-section-2"><span class="toc-item-num">2  </span>Second main section</a></span><ul class="toc-item"><li><span><a href="#subsection-c" data-toc-modified-id="subsection-c-2.1"><span class="toc-item-num">2.1  </span>subsection c</a></span></li><li><span><a href="#subsection-d" data-toc-modified-id="subsection-d-2.2"><span class="toc-item-num">2.2  </span>subsection d</a></span></li></ul></li></ul></div>
```

**Comment by jcb91** [comment]
plus my versions:

```
$ jupyter notebook --version
5.0.0

$ jupyter contrib nbextensions --version
0.3.3

$ jupyter nbconvert --version
5.3.1
```

**Comment by jfbercher** [comment]
@p2ya, @jcb91 I am also unable to reproduce with notebook 5.0, nbconvert 5.2 and contribs 0.3 + various personal patches. After upgrading everything 5.2/5.3.1/0.3.3 still unable. Anyway, I don't see how this can happen.. 

@jcb91 I don't see (dont't remember) also how the `toc_cell` parameter could do anything on the non live notebook. The `toc_cell` is added statically with links embedded in. #1141 is elegant but does not change the scene I think.

**Comment by p2ya** [comment]
my versions:

$ jupyter notebook --version
5.1.0

$ jupyter contrib nbextensions --version
0.3.1

$ jupyter nbconvert --version
5.3.1

**Comment by p2ya** [comment]
I just created a markdown notebook with side and cell toc.
the issue is still valid when I converted to the html.
Can you please check if you get the same.

[tocTest.ipynb.zip](https://github.com/ipython-contrib/jupyter_contrib_nbextensions/files/1416555/tocTest.ipynb.zip)

**Comment by jfbercher** [comment]
Indeed, the issue you mention is present (the links in toc_cell show localhost:8889 which are not relative).
Please update to jupyter_contrib_nbextensions 0.3.3 and test again (delete toc_cell and recreate it) before investigating further.

**Comment by jcb91** [comment]
> @jcb91 I don't see (dont't remember) also how the `toc_cell` parameter could do anything on the non live notebook. The `toc_cell` is added statically with links embedded in. #1141 is elegant but does not change the scene I think.

Sort of. Although added statically, it is (since f7059e07074600373b6dc0e170afcf24f23b7dc0) also updated on loading of a static notebook. But anyway, I was fishing for ways the rewriting might not be happening, but as is now clear, the rewriting isn't necessary at all, if the original ToC cell construction has gone correctly.

**Comment by p2ya** [comment]
the upgrade worked nicely, thanks for your help

**Comment by jfbercher** [comment]
@p2ya  Nice to read that it is solved!

@jcb91 
> the rewriting [of toc_cell] isn't necessary at all, if the original ToC cell construction has gone correctly.

Yes. it is useless and we should remove it 
(or we may insert the toc_cell in a div say with class toc_cell and enable to populate it even in the non-live notebook, using the "t" or the new incoming option in option window, even if the toc_cell isn't present at the beginning)

**Comment by jcb91** [comment]
> Yes. it is useless and we should remove it

Not quite useless, since it attaches the specialised click handler which uses the `data-toc-modified-id` attributes instead of just the `id`s. Although it doesn't quite work correctly right now, as I've messed up the jquery selectors.

> or we may insert the toc_cell in a div say with class toc_cell and enable to populate it even in the non-live notebook, using the "t" or the new incoming option in option window, even if the toc_cell isn't present at the beginning

We could, though it's not trivial. Locating and repopulating an existing toc_cell in a non-live notebook can already be accomplished using, for example

```javascript
$('.cell .toc').parent().children('h1').children('.tocSkip').closest('.cell')
```

to find the `.cell` that contains (a `.toc` which has (an `h1` that contains a `.tocSkip`) as a sibling). So far it isn't removed if `cfg.cell_toc` is `false` (which without #1141 or similar it always will be), there's no way to alter the config value in a non-live notebook (though that'd be easy to add on the back of #1124), and we'd have to figure out how to construct the cell from scratch. I just went with only doing the rerendering as a simple halfway, offering a small benefit without much effort (and messed it up, though thankfully in a quiet way!)...
